x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1968,28 +2117,21 @@ String prependVersionAndEncodeAccessToken(TransportVersion version, String acces
         }
     }
 
-    public static String prependVersionAndEncodeRefreshToken(TransportVersion version, String payload) {
-        try (BytesStreamOutput out = new BytesStreamOutput()) {
-            out.setTransportVersion(version);
-            TransportVersion.writeVersion(version, out);
-            out.writeString(payload);
-            return Base64.getEncoder().encodeToString(out.bytes().toBytesRef().bytes);
-        } catch (IOException e) {
-            throw new RuntimeException("Unexpected exception when working with small in-memory streams", e);
-        }
-    }
-
-    // public for testing
-    /**
-     * Unpacks a base64 encoded pair of a version tag and String payload.
-     */
-    public static Tuple<TransportVersion, String> unpackVersionAndPayload(String encodedPack) throws IOException {
-        final byte[] bytes = encodedPack.getBytes(StandardCharsets.UTF_8);
-        try (StreamInput in = new InputStreamStreamInput(Base64.getDecoder().wrap(new ByteArrayInputStream(bytes)), bytes.length)) {
-            final TransportVersion version = TransportVersion.readVersion(in);
-            in.setTransportVersion(version);
-            final String payload = in.readString();
-            return new Tuple<>(version, payload);
+    public static String prependVersionAndEncodeRefreshToken(TransportVersion version, byte[] refreshTokenBytes) throws IOException {
+        if (version.onOrAfter(VERSION_GET_TOKEN_DOC_FOR_REFRESH)) {
+            try (BytesStreamOutput out = new BytesStreamOutput()) {
+                out.setTransportVersion(version);
+                TransportVersion.writeVersion(version, out);
+                out.writeByteArray(refreshTokenBytes);
+                return Base64.getEncoder().encodeToString(out.bytes().toBytesRef().bytes);
+            }
+        } else {
+            try (BytesStreamOutput out = new BytesStreamOutput()) {
+                out.setTransportVersion(version);
+                TransportVersion.writeVersion(version, out);
+                out.writeString(Base64.getUrlEncoder().withoutPadding().encodeToString(refreshTokenBytes));
+                return Base64.getEncoder().encodeToString(out.bytes().toBytesRef().bytes);
```

Comment:
Yeah, I agree the usage of both `Base64.getUrlEncoder()` and `Base64.getEncoder()` is confusing and a possible source of bugs...
So, I think I can make base64 encoding uniform (use the non-URL-safe one), at least for post 7.2 tokens (I plan to remove pre 7.2 support altogether in a follow-up) though this slightly exceeds the scope of this PR, I will help a lot with readability...
Let me know if you're against fixing it in this PR.

Reply:
Thanks for explanation. I'm fine either way. Do what's easier for you.

Reply:
I was wrong. It's not possible to converge the base64 encoding schemes.... I can only change the encoding used in the new token version. The old tokens, as well as the encrypted superseding tokens, must continue to use whatever encoding/decoding scheme was already in place.
Is it worth changing regular base64 encoding over the url-safe one for new tokens only? I'm oscillating...

Reply:
🤔 Looking at the RFCs ([6750](https://datatracker.ietf.org/doc/html/rfc6750#section-2.1) and updated [9110](https://datatracker.ietf.org/doc/html/rfc9110#name-authentication-parameters)) they don't state that the token has to be `base64` nor `base64Url` encoded. It just states that it has to support characters which allow usage of `base64` and `base64Url` encoded tokens. So I think we are fine in both cases. I'll defer to you for the decision.

Reply:
Tokens returned to users are always `base64` encoded (see eg https://github.com/elastic/elasticsearch/blob/b504346ff4bb76b81087ea39fbcd9c59af4a2889/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L1938). The issue here is that stuff that gets stored in token docs sometimes uses `base64` and sometimes `base64Url`. And since we can't get to a uniform convention because of bwc concerns, I'm leaning to leave it as is.